agent/src/main.rs

```diff
@@ -580,7 +613,19 @@ fn apply_smf(
 
     for (_, region_snapshots) in running_snapshots.iter_mut() {
         for snapshot in region_snapshots.values_mut() {
-            if snapshot.state != State::Requested {
+            // The agent is not responsible for creating snapshots, only
+            // starting a read-only downstairs that points to them. When the
+            // original request to start a running snapshot is made, the object
+            // begins in state Requested, and once apply_smf returns
+            // successfully has its state changed to Created (this is used by
+            // the caller to query if the service has been created yet).
+            //
+            // If the agent zone is bounced, we want it to read the datafile
+            // and restore all the previously created SMF instances. This means
+            // that we have to take action on both Requested and Created for
+            // running snapshots. This is in contrast to Region, which has
+            // different actions for Requested and Created.
```

Comment:
Is the agent single threaded here?  Like could two different tasks/threads both call apply_smf_actual()?  I just want to be sure that the startup case, where we have a bunch of restart work to do won't clash with some external entity trying to act on the same snapshot at the same time, and being confused for the startup path.

Reply:
As far as I can tell it runs two threads: one for the worker and one for the dropshot server:

```rust
            let log0 = log.new(o!("component" => "worker"));
            let df0 = Arc::clone(&df);
            std::thread::spawn(|| {
                worker(
                    log0,
                    df0,
                    regions_dataset,
                    downstairs_program,
                    downstairs_prefix,
                    snapshot_prefix,
                )
            });

            server::run_server(&log, listen, df).await
```

datafile uses interior mutability to make changes to the datafile, and both those threads share an `Arc<Datafile>`. It's possible for state changes to interleave because neither the server nor the worker hold a persistent guard to the inner part of the datafile. I don't think there's any guarding against multiple threads calling `apply_smf`, but only the worker function calls it: the server thread only makes datafile changes then signals for the worker via the Condvar.

Reply:
Note that the agent does call `apply_smf` when starting up, before spawning the worker and starting the dropshot server.